aten/src/ATen/native/cuda/CUDALoops.cuh

```diff
@@ -267,23 +169,28 @@ __device__ inline void elementwise_kernel_helper(func_t f, array_t data, policy_
   }
 
   // store
-  auto result_accessor = [&] __device__ (int index) -> return_t & { return results[index]; };
-  policy.store(result_accessor, result_base);
+  policy.store(results, idx);
 }
 
 template<int vec_size, typename func_t, typename array_t>
 C10_LAUNCH_BOUNDS_1(num_threads)
-__global__ void elementwise_kernel(int N, func_t f, array_t data) {
-  using return_t = typename function_traits<func_t>::result_type;
+__global__ void vectorized_elementwise_kernel(int N, func_t f, array_t data) {
   int remaining = N - block_work_size * blockIdx.x;
 
   if (remaining < block_work_size) {  // if this block handles the reminder, just do a naive unrolled loop
-    elementwise_kernel_helper(f, data, typename memory::policies::checked_unroll(remaining));
+    elementwise_kernel_helper(f, typename memory::policies::unroll<array_t>(data, remaining));
   } else {  // if this block has a full `block_work_size` data to handle, use vectorized memory access
-    elementwise_kernel_helper(f, data, typename memory::policies::template vectorized<vec_size>());
+    elementwise_kernel_helper(f, typename memory::policies::template vectorized<vec_size, array_t>(data));
   }
 }
 
+template<typename func_t, typename array_t>
+C10_LAUNCH_BOUNDS_1(num_threads)
+__global__ void unrolled_elementwise_kernel(int N, func_t f, array_t data) {
```

Comment:
This kernel will be used to handle all other cases: non-contiguous, and dynamic casting